diff --git a/ds4drv/config.py b/ds4drv/config.py
index d01a791..c0f61d1 100644
--- a/ds4drv/config.py
+++ b/ds4drv/config.py
@@ -69,7 +69,7 @@ daemonopt.add_argument("--daemon-pid", default=DAEMON_PID_FILE, metavar="file",
 controllopt = parser.add_argument_group("controller options")
 
 
-class Config(configparser.SafeConfigParser):
+class Config(configparser.ConfigParser):
     def load(self, filename):
         self.read([filename])
 
@@ -96,7 +96,7 @@ class Config(configparser.SafeConfigParser):
             return {}
 
     def sections(self, prefix=None):
-        for section in configparser.SafeConfigParser.sections(self):
+        for section in configparser.ConfigParser.sections(self):
             match = re.match(r"{0}:(.+)".format(prefix), section)
             if match:
                 yield match.group(1), section
