diff --git a/oblogout/__init__.py b/oblogout/__init__.py
index 2202739..74115e5 100755
--- a/oblogout/__init__.py
+++ b/oblogout/__init__.py
@@ -190,7 +190,7 @@ class OpenboxLogout():
         """ Load the configuration file and parse entries, when encountering a issue
             change safe defaults """
 
-        self.parser = configparser.SafeConfigParser()
+        self.parser = configparser.ConfigParser()
         self.parser.read(config)
 
         # Set some safe defaults
