diff --git a/_scripts/build.js b/_scripts/build.js
index 035f986c..70c456e4 100644
--- a/_scripts/build.js
+++ b/_scripts/build.js
@@ -36,12 +36,13 @@ if (platform === 'darwin') {
     arch = Arch.armv7l
   }
 
-  targets = Platform.LINUX.createTarget(['deb', 'zip', '7z', 'apk', 'rpm', 'AppImage', 'pacman'], arch)
+  targets = Platform.LINUX.createTarget(['dir'], arch)
 }
 
 const config = {
   appId: `io.freetubeapp.${name}`,
   copyright: 'Copyleft © 2020-2023 freetubeapp@protonmail.com',
+  electronDist: '/usr/lib/electron24',
   // asar: false,
   // compression: 'store',
   productName,
