diff -ruN a/data/lightdm-gtk-greeter.conf b/data/lightdm-gtk-greeter.conf
--- a/data/lightdm-gtk-greeter.conf	2020-06-11 20:00:31.000000000 -0400
+++ b/data/lightdm-gtk-greeter.conf	2023-11-11 21:53:20.375043706 -0500
@@ -48,11 +48,12 @@
 #  transition-duration = overrides default value
 #
 [greeter]
-#background=
-#user-background=
-#theme-name=
-#icon-theme-name=
-#font-name=
+background = /usr/share/backgrounds/wallpaper1.png
+user-background = false
+theme-name = Graphite-black-green-Dark
+icon-theme-name = Tela-green-dark
+font-name = Roboto 10
+cursor-theme-name = Graphite-light-cursors
 #xft-antialias=
 #xft-dpi=
 #xft-hintstyle=
@@ -61,5 +62,5 @@
 #clock-format=
 #keyboard=
 #reader=
-#position=
+position =  35%,end 50%,center
 #screensaver-timeout=
