diff --git a/src/modules/localecfg/main.py b/src/modules/localecfg/main.py
index 8f0edcd17..bc2d32297 100644
--- a/src/modules/localecfg/main.py
+++ b/src/modules/localecfg/main.py
@@ -1,4 +1,4 @@
-#!/usr/bin/env python3
+#!/usr/bin/python3
 # -*- coding: utf-8 -*-
 #
 # === This file is part of Calamares - <https://calamares.io> ===
@@ -45,7 +45,7 @@ def extract_locale(line):
     Extracts a locale from the @p line, and returns a pair of
     (extracted-locale, uncommented line). The locale is the
     first word of the line after uncommenting (in the human-
-    readable text explanation at the top of most /etc/locale.gen
+    readable text explanation at the top of most /etc/default/libc-locales
     files, the locales may be bogus -- either "" or e.g. "Configuration")
     """
     # Remove leading spaces and comment signs
@@ -63,7 +63,7 @@ def extract_locale(line):
 
 def rewrite_locale_gen(srcfilename, destfilename, locale_conf):
     """
-    Copies a locale.gen file from @p srcfilename to @p destfilename
+    Copies a libc-locales file from @p srcfilename to @p destfilename
     (this may be the same name), enabling those locales that can
     be found in the map @p locale_conf. Also always enables en_US.UTF-8.
     """
@@ -135,7 +135,7 @@ def run():
         return (_("Configuration Error"),
                 _("No root mount point is given for <pre>{!s}</pre> to use." ).format("localecfg"))
 
-    target_locale_gen = "{!s}/etc/locale.gen".format(install_path)
+    target_locale_gen = "{!s}/etc/default/libc-locales".format(install_path)
     target_locale_gen_bak = target_locale_gen + ".bak"
     target_locale_conf_path = "{!s}/etc/locale.conf".format(install_path)
     target_etc_default_path = "{!s}/etc/default".format(install_path)
@@ -147,11 +147,11 @@ def run():
             .format(target_locale_gen_bak, target_locale_gen))
 
     # run locale-gen if detected; this *will* cause an exception
-    # if the live system has locale.gen, but the target does not:
+    # if the live system has libc-locales, but the target does not:
     # in that case, fix your installation filesystem.
-    if os.path.exists('/etc/locale.gen'):
+    if os.path.exists('/etc/default/libc-locales'):
         rewrite_locale_gen(target_locale_gen, target_locale_gen, locale_conf)
-        libcalamares.utils.target_env_call(['locale-gen'])
+        libcalamares.utils.target_env_call(['xbps-reconfigure', '-f', 'glibc-locales'])
         libcalamares.utils.debug('{!s} done'.format(target_locale_gen))
 
     # write /etc/locale.conf
